lib/libc/gen/emutls.c

```diff
@@ -0,0 +1,311 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Andrew Turner
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory under DARPA/AFRL contract FA8750-10-C-0237
+ * ("CTSRD"), as part of the DARPA CRASH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/param.h>
+
+#include <machine/cpu.h>
+
+#include <pthread.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef EMUTLS_LIBC
+#include "static_tls.h"
+#include "libc_private.h"
+#endif
+
+struct emutls_control {
+	uint64_t size;
+	uint64_t align;
+	union {
+		_Atomic size_t index;
+		void *unused; /* Ensure this is pointer sized */
+	};
+	void *value_addr;
+};
+
+_Static_assert(sizeof(struct emutls_control) == (sizeof(uint64_t) * 2 +
+    sizeof(void *) * 2), "emutls_control is the wrong size");
+
+struct emutls_array {
+	size_t size;
+	void *data[];
+};
+
+#ifdef EMUTLS_LIBC
+struct emutls_array_list {
+	struct emutls_array_list *next;
+	struct emutls_array *array;
+	uintptr_t handle;
+};
+
+/* Hack to not need to call pthread code that may use an allocator that then
+ * used tls data that calls into this code. The address of this is the handle
+ * to identify the thread. This means this file must not be built with emulated
+ * tls enabled.
+ */
```

Comment:
This comment looks stale.

Reply:
I think it's still true, but probably it should be moved before the struct and reworded slightly as a description of the EMUTLS_LIBC case.